make.sh

```diff
@@ -0,0 +1,182 @@
+ -h|--help           Print this message and exit.
+"
+    exit 1
+}
+
+function check_keys {
+    if ! [ -e "$SCRIPT_DIR/misc/private-key-OUT.pem" ]; then
+        return 1
+    fi
+
+    return 0
+}
+
+function remove_keys {
+    echo "Removing old keys!"
+
+    if [ -e "$MISC_DIR/private-key-OUT.pem" ]; then
+        echo "Removing private key"
+        rm "$MISC_DIR/private-key-OUT.pem"
+    fi
+
+    if [ -e "$MISC_DIR/wwfcPayloadPublicKey-OUT.hpp" ]; then
+        echo "Removing public key"
+        rm "$MISC_DIR/wwfcPayloadPublicKey-OUT.hpp"
+    fi
+}
+
+if [ $# -eq 0 ]; then
+    >&2 echo "No arguments provided"
+    print_help
+fi
+
+clean=false
+clean_keys=false
+try_make_keys=false
+keys=false
+patch=false
+stage1=false
+payload=false
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+    --clean)
+        clean=true
+        ;;
+    --clean-keys)
+        clean_keys=true
+        ;;
+    --all)
+        patch=true
+        stage1=true
+        payload=true
+        try_make_keys=true
+        ;;
+    --keys)
+        keys=true
+        ;;
+    --patch)
+        patch=true
+        ;;
+    --stage1)
+        stage1=true
+        ;;
+    --payload)
+        payload=true
+        ;;
+    -h | --help)
+        print_help
+        ;;
+    *)
+        echo -e "Bad option: $1\n"
+        print_help
+        ;;
+    esac
+    shift
+done
+
+if $clean; then
+    if [ -d "$DIST_DIR" ]; then
+        echo "Removing dist"
+        rm -r "$DIST_DIR"
+    fi
+
+    if [ -d "$PATCH_DIR/build" ]; then
+        echo "Cleaning $PATCH_DIR"
+        rm -r "$PATCH_DIR/build"
+    fi
+
+    if [ -d "$STAGE1_DIR/build" ]; then
+        echo "Cleaning $STAGE1_DIR"
+        rm -r "$STAGE1_DIR/build"
+    fi
+
+    if [ -d "$PAYLOAD_DIR/build" ]; then
+        echo "Removing $PAYLOAD_DIR/build"
+        rm -r "$PAYLOAD_DIR/build"
+    fi
+
+    if [ -d "$PAYLOAD_DIR/binary" ]; then
+        echo "Removing $PAYLOAD_DIR/binary"
+        rm -r "$PAYLOAD_DIR/binary"
+    fi
+fi
+
+if $clean_keys; then
+    remove_keys
+fi
+
+mkdir -p "$DIST_DIR"
+
+if $keys || ( (! check_keys) && $try_make_keys); then
+    remove_keys
+
+    cd "$MISC_DIR"
+    echo "Creating new keys!"
+
+    python generate-key.py new private-key-OUT.pem wwfcPayloadPublicKey-OUT.hpp
+
+    echo "Copying wwfcPayloadPublicKey-OUT.hpp to $SCRIPT_DIR/include/wwfcPayloadPublicKey.hpp"
+    cp "$MISC_DIR/wwfcPayloadPublicKey-OUT.hpp" "$SCRIPT_DIR/include/wwfcPayloadPublicKey.hpp"
+    echo "Copying private-key-OUT.pem to dist"
+    cp "$MISC_DIR/private-key-OUT.pem" "$DIST_DIR/private-key.pem"
+fi
+
+if $payload || $stage1 || $patch; then
+    if ! check_keys; then
+        echo "$SCRIPT_DIR/misc/private-key-OUT.pem not found! Please generate keys using '--keys' before attempting to build anything else"
+        exit 1
+    fi
+fi
+
+if $payload; then
+    cd "$PAYLOAD_DIR"
+    echo "Making payloads!"
+
+    python make-payload.py -j16
```

Comment:
Probably should make the number of cores used as an argument in the script rather than hardcode 16.

Reply:
Good catch, fixed by allowing additional args for each script to be provided after `--`